File: leetcode/LowestCommonAncestor.java
class Solution {
    
    private TreeNode lca; 
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        
        if (root == null) {
            return null; 
        }
      
        TreeNode left = root; 
        TreeNode right = root;     
        
        
        while (left != null && right != null) {
            
            if (left == right ) 
                lca = left; 
            
            if (left == p && right == q) {
                break;
            }
            
            if (p.val < left.val) {
                left = left.left ;
            } else if (p.val > left.val ) {
                left = left.right; 
            } 
            
            if (q.val < right.val) {
                right = right.left;
            } else if (q.val > right.val) {
                right = right.right; 
            }

        }
        
        return lca; 
    }
}

File: leetcode/lowestCommonAncestor.py
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        if root is None:
            return None
        
        if root == p or root == q:
            return root 
        
        left = self.lowestCommonAncestor(root.left, p, q)
        
        right = self.lowestCommonAncestor(root.right, p, q)
        
        if left is not None and right is not None:
            return root
        
        if left is None and right is None:
            return None
        
        return left if left is not None else right